base_layer/core/src/validation/block_body/block_body_full_validator.rs

```diff
@@ -94,7 +94,25 @@ impl BlockBodyFullValidator {
             );
             ChainStorageError::AccessError("write lock on smt".into())
         })?;
-        let mmr_roots = chain_storage::calculate_mmr_roots(backend, &self.consensus_manager, &block, &mut output_smt)?;
+        let mmr_roots =
+            match chain_storage::calculate_mmr_roots(backend, &self.consensus_manager, &block, &mut output_smt) {
+                Ok(mmr_roots) => mmr_roots,
+                Err(e) => {
+                    error!(
+                        target: LOG_TARGET,
+                        "Validator could not calculate MMR roots for block {}: {:?}", block.hash().to_hex(), e
+                    );
+                    match e {
+                        ChainStorageError::CannotCalculateNonTipMmr(ref _e) => {
+                            return Err(e.into());
+                        },
+                        _ => {},
+                    }
+                    // Recalculate SMT as it might have been altered.
+                    *output_smt = backend.calculate_tip_smt()?;
```

Comment:
To re-calculate the smt inside a validator seems out of place - validate should only validate and not fix things

Reply:
This is due to the bad design of the SMT. It can get corrupted in calculate_mmr_roots and this is to roll it back on error. I have not added logic here, just matched the logic that is **present in other places** but missing here. 

Background: calculate_mmr_roots changes the SMT, then calculates the root, then undoes changes to the SMT. If there is an error during `calculate_mmr_roots` and it exits early we need to revert the SMT.